# Wikipedia Notable Life Expectancies

# [Notebook 3 of 4: Data Cleaning](https://github.com/teresahanak/wikipedia-life-expectancy/blob/main/wp_life_expect_data_clean2_thanak_2022_06_17.ipynb)

## Context

The


## Objective

The

### Data Dictionary

- Feature: Description

## Importing Necessary Libraries

In [1]:
# To structure code automatically
%load_ext nb_black

# To import/export sqlite databases
import sqlite3 as sql

# To help with reading and manipulating data
import pandas as pd
import numpy as np
import re

# To define maximum number of columns to be displayed in a dataframe
pd.set_option("display.max_columns", None)
# To define the maximum number of rows to be displayed in a dataframe
pd.set_option("display.max_rows", 200)

# To supress warnings
# import warnings

# warnings.filterwarnings("ignore")

# To set some visualization attributes
pd.set_option("max_colwidth", 150)

<IPython.core.display.Javascript object>

## Data Overview


### Reading, Sampling, and Checking Data Shape

In [2]:
# Reading the dataset
conn = sql.connect("wp_life_expect_clean1.db")
data = pd.read_sql("SELECT * FROM wp_life_expect_clean1", conn)

# Making a working copy
df = data.copy()

# Checking the shape
print(f"There are {df.shape[0]} rows and {df.shape[1]} columns.")

# Checking first 2 rows of the data
df.head(2)

There are 132445 rows and 20 columns.


day              name                                                 info  \
0   1  William Chappell  , 86, British dancer, ballet designer and director.   
1   1    Raymond Crotty         , 68, Irish economist, writer, and academic.   

                                                      link num_references  \
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21   
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12   

   year    month info_parenth info_1           info_2  \
0  1994  January         None          British dancer   
1  1994  January         None         Irish economist   

                         info_3        info_4 info_5 info_6 info_7 info_8  \
0  ballet designer and director          None   None   None   None   None   
1                        writer  and academic   None   None   None   None   

  info_9 info_10 info_11   age  
0   None    None    None  86.0  
1   None    None    None  68.0

<IPython.core.display.Javascript object>

In [3]:
# Checking last 2 rows of the data
df.tail(2)

day           name  \
132443   9  Oleg Moliboga   
132444   9       Zou Jing   

                                                                         info  \
132443          , 69, Russian volleyball player, Olympic champion  and coach.   
132444  , 86, Chinese engineer, member of the Chinese Academy of Engineering.   

                                                     link num_references  \
132443        https://en.wikipedia.org/wiki/Oleg_Moliboga              2   
132444  https://en.wikipedia.org/wiki/Zou_Jing_(engineer)              3   

        year month info_parenth info_1                     info_2  \
132443  2022  June           ()         Russian volleyball player   
132444  2022  June         None                  Chinese engineer   

                                              info_3 info_4 info_5 info_6  \
132443                   Olympic champion  and coach   None   None   None   
132444  member of the Chinese Academy of Engineering   None   None   None   

       info_7 info_8 info_9 info_10 info_11   age  
132443   None   None   None    None    None  69.0  
132444   None   None   None    None    None  86.0

<IPython.core.display.Javascript object>

In [4]:
# Checking a sample of the data
df.sample(5)

day                       name  \
105933   8            Eberhard Ferstl   
43418   22          Sir Percy Cradock   
104218  11                Neil Estern   
693     15                 John Curry   
88269   16  Satrio Budihardjo Joedono   

                                                               info  \
105933  , 86, German Olympic bronze medallist field hockey player .   
43418                  , 86, British diplomat, after short illness.   
104218                                     , 93, American sculptor.   
693                      , 44, British figure skater, heart attack.   
88269     , 84, Indonesian economist, Chairman of the Audit Board .   

                                                           link  \
105933            https://en.wikipedia.org/wiki/Eberhard_Ferstl   
43418               https://en.wikipedia.org/wiki/Percy_Cradock   
104218                https://en.wikipedia.org/wiki/Neil_Estern   
693                    https://en.wikipedia.org/wiki/John_Curry   
88269   https://en.wikipedia.org/wiki/Satrio_Budihardjo_Joedono   

       num_references  year    month info_parenth info_1  \
105933              2  2019  October           ()          
43418              35  2010  January         None          
104218             25  2019     July         None          
693                28  1994    April         None          
88269               0  2017    April          (–)          

                                                     info_2  \
105933  German Olympic bronze medallist field hockey player   
43418                                      British diplomat   
104218                                    American sculptor   
693                                   British figure skater   
88269                                  Indonesian economist   

                             info_3 info_4 info_5 info_6 info_7 info_8 info_9  \
105933                         None   None   None   None   None   None   None   
43418           after short illness   None   None   None   None   None   None   
104218                         None   None   None   None   None   None   None   
693                    heart attack   None   None   None   None   None   None   
88269   Chairman of the Audit Board   None   None   None   None   None   None   

       info_10 info_11   age  
105933    None    None  86.0  
43418     None    None  86.0  
104218    None    None  93.0  
693       None    None  44.0  
88269     None    None  84.0

<IPython.core.display.Javascript object>

#### Observations:
- There are currently 132,445 rows and 20 columns.

### Checking Data Types, Duplicates, and Null Values

In [5]:
# Checking data types and null values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132445 entries, 0 to 132444
Data columns (total 20 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   day             132445 non-null  object 
 1   name            132445 non-null  object 
 2   info            132445 non-null  object 
 3   link            132445 non-null  object 
 4   num_references  132445 non-null  object 
 5   year            132445 non-null  int64  
 6   month           132445 non-null  object 
 7   info_parenth    49788 non-null   object 
 8   info_1          132445 non-null  object 
 9   info_2          132421 non-null  object 
 10  info_3          62570 non-null   object 
 11  info_4          12587 non-null   object 
 12  info_5          1505 non-null    object 
 13  info_6          217 non-null     object 
 14  info_7          33 non-null      object 
 15  info_8          7 non-null       object 
 16  info_9          1 non-null       object 
 17  info_10   

<IPython.core.display.Javascript object>

#### Observations:
- Our dataset was saved to and read from the database without any hiccups.
- Picking up where we left off, we will aim to track down the remaining missing values for `age`, starting with searching for digits in `info_2`.

### Remaining Missing Values for Age

In [6]:
# Checking number of remaining missing values
print(f'There are {df["age"].isna().sum()} missing values for age.')

There are 79 missing values for age.


<IPython.core.display.Javascript object>

#### Function to Save Indices of Rows Matching Regular Expressions Pattern to a List and Print Number of Rows with Match

In [7]:
# Define a function that takes dataframe, column name, and re pattern as arguments and returns list of indices
# for which column value matches re pattern
def rows_with_pattern(dataframe, column, pattern):
    """
    Takes input of dataframe, column name, and re pattern 
    and returns list of indices for rows that contain match
    for pattern anywhere within value for given column.
    
    dataframe: dataframe
    column: column name
    pattern: re pattern
    """
    index_list = []

    for i in dataframe.index:
        item = dataframe.loc[i, column]
        match = re.search(pattern, item)
        if match:
            index_list.append(i)
    print(
        f"There are {len(index_list)} rows with matching pattern in column '{column}'."
    )
    return index_list

<IPython.core.display.Javascript object>

#### Function to Use rows_with_pattern Function for Multiple Regular Expression Patterns

In [8]:
# Define a function that calls rows_with_pattern function for multiple re patterns
# returning a single list of indices for all rows with any pattern match


def multiple_patterns(dataframe, column, patterns):
    """
    Takes input dataframe, column, and list of re patterns and returns single list 
    of indices for rows in which a match for any pattern is found with re.search
    
    dataframe: dataframe
    column: column name
    patterns: list of re patterns
    """
    rows_combined = []

    # For loop to check each pattern
    for pattern in patterns:

        # List and number of rows matching each pattern
        print(pattern)
        rows_to_check = rows_with_pattern(dataframe, column, pattern)
        print("")

        # Add list for each pattern to combined list
        rows_combined += rows_to_check

    return rows_combined

<IPython.core.display.Javascript object>

### `info_2`

#### Rows Missing `age` with Digits in `info_2`

In [9]:
# Pattern for re
pattern = r"\d"

# List and number of rows matching pattern
rows_to_check = rows_with_pattern(
    df[(df["age"].isna()) & (df["info_2"].notna())], "info_2", pattern
)

# Examining the rows directly
df.loc[rows_to_check, :]

There are 43 rows with matching pattern in column 'info_2'.


day                            name  \
6846    17                     Spiro Agnew   
12861   14                     Muslimgauze   
14993   15               Željko Ražnatović   
15062   28                  Sarah Caudwell   
16111   30                   Max Showalter   
17792   15                     Joey Ramone   
18170    4                        Dipendra   
19017   28          Mohammad Khalequzzaman   
19117   12  Lord Hailsham of St Marylebone   
23752    6                     Jules Engel   
24882    8                 Muhammad Zaidan   
25173   30         Kioumars Saberi Foumani   
25201    6                  Kjell Hallbing   
26020   29               Shimon Wincelberg   
26819   28                  Karen Lancaume   
26978   16                  Narriman Sadek   
26996   19                 Kihachi Okamoto   
27024   22                      Lee Eun-ju   
27317    2               Pope John Paul II   
27726   30               Takanohana Kenshi   
28376   28                 Esther Szekeres   
28765   22                           Arman   
28961   15                Agenore Incrocci   
30074   20                  Stanley Hiller   
32145    9                     Cocoa Samoa   
61916    1                     Basil Soper   
86908    8              Mohamud Muse Hersi   
110601  12              Khalif Mumin Tohow   
110654  14                   Danny Delaney   
111267   2                   Justa Barrios   
111798  21              Kamrun Nahar Putul   
112645  23    Jean-Michel Bokamba-Yangouma   
113244  16             Cornelius Mwalwanda   
115179   2                       Fadma Abi   
121298  18           Michael Bedford-Jones   
121419  21                       Johny Lal   
122041  10           Salem Chandrasekharan   
122362  20                U. Visweswar Rao   
122634  30                    Venkat Subha   
123181  21                     Haribhushan   
124505   8        Fernando López de Olmedo   
126767   2                     Irene Lalji   
127156  17                R. N. R. Manohar   

                                                                                                          info  \
6846                            , American politician, 77, 39th Vice President of the United States, leukemia.   
12861                                                   , , 37, British electronic musician, fungal infection.   
14993                                                         , , 47, Serbian mobster and paramilitary leader.   
15062                                            , , 60, British detective story writer and barrister, cancer.   
16111                                               , , 83, American actor, composer, pianist, singer, cancer.   
17792                                        , , 49, American musician, lead singer for The Ramones, lymphoma.   
18170                                                                            , King of Nepal, 29, suicide.   
19017   , member of the then National Assembly of Pakistan and Union Minister of Labor, died in 28 September .   
19117                                                                   , , 94, British lawyer and politician.   
23752                                 , Jules Engel, 94, American filmmaker, visual artist, and film director.   
24882                                  , , 55, Palestinian nationalist, founder of Palestine Liberation Front.   
25173                                                                                , , 62, Iranian satirist.   
25201                                                                        , , 69, Norwegian Western author.   
26020                                                                      , , 80, American television writer.   
26819                                        , , 32, French adult film performer, overdosed on sleeping pills.   
26978                                     , , 70, Egyptian queen, ex-wife of King Farouk, last queen of Egypt.   
26996                                       

<IPython.core.display.Javascript object>

#### Observations:
- There are several ages as sole integer year values and two as year ranges with two integers.
- The remaining entries are missing age, but do have digits, so order of processing matters here.
- We can safely remove any of these rows that contains a letter of the alphabet, taking care to select rows only from those that are missing `age` and have a digit in `info_2`.

In [10]:
# Pattern for re
pattern = r"[a-z,A-Z]"

# List and number of rows matching pattern
rows_to_check = rows_with_pattern(df.loc[rows_to_check, :], "info_2", pattern)

# Checking a sample of the rows
df.loc[rows_to_check, :].sample(2)

There are 17 rows with matching pattern in column 'info_2'.


day                      name  \
124505   8  Fernando López de Olmedo   
113244  16       Cornelius Mwalwanda   

                                                    info  \
124505  , Spanish general commander of Ceuta , COVID-19.   
113244    , Malawian economist and politician, COVID-19.   

                                                               link  \
124505  https://en.wikipedia.org/wiki/Fernando_L%C3%B3pez_de_Olmedo   
113244            https://en.wikipedia.org/wiki/Cornelius_Mwalwanda   

       num_references  year   month             info_parenth  \
124505              4  2021  August  (Perejil Island crisis)   
113244              5  2020    July                     None   

                                    info_1    info_2 info_3 info_4 info_5  \
124505  Spanish general commander of Ceuta  COVID-19   None   None   None   
113244   Malawian economist and politician  COVID-19   None   None   None   

       info_6 info_7 info_8 info_9 info_10 info_11  age  
124505   None   None   None   None    None    None  NaN  
113244   None   None   None   None    None    None  NaN

<IPython.core.display.Javascript object>

#### Observations:
- We can drop these rows, as they are missing the data for age.
- Extraction of age for single integer value and two integer ranges can follow.

#### Dropping Additional Rows with Age Data Absent

In [11]:
# Dropping rows, resetting index, and checking new shape of df
df.drop(rows_to_check, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132428, 20)

<IPython.core.display.Javascript object>

#### Remaining Rows with `age` values in `info_2`

In [12]:
# Regular expression for parenthesis and its contents
pattern = r"\d"

# Finding indices of rows that have pattern
rows_to_check = rows_with_pattern(
    df[(df["age"].isna()) & (df["info_2"].notna())], "info_2", pattern
)

# Checking unique values
df.loc[rows_to_check, :]["info_2"].unique()

There are 26 rows with matching pattern in column 'info_2'.


array(['77', '37', '47', '60', '83', '49', '29', '94', '55', '62', '69',
       '80', '32', '70', '81', '24', '84', '95', '76', '86', '61',
       '74–75', '79–80'], dtype=object)

<IPython.core.display.Javascript object>

#### Extracting `age` for Ranges with Two Values

In [13]:
# Pattern for re
pattern = r"(\d{1,3})(-|–|/| or )(\d{1,3})"

# List and number of rows matching pattern
rows_to_check = rows_with_pattern(
    df[(df["age"].isna()) & (df["info_2"].notna())], "info_2", pattern
)

# Checking sample of rows
df.loc[rows_to_check, :].sample(2)

There are 2 rows with matching pattern in column 'info_2'.


day                name  \
61915   1         Basil Soper   
86907   8  Mohamud Muse Hersi   

                                                      info  \
61915                              , British actor, 74–75.   
86907  , Somali politician, 79–80, President of Puntland .   

                                                   link num_references  year  \
61915         https://en.wikipedia.org/wiki/Basil_Soper              0  2013   
86907  https://en.wikipedia.org/wiki/Mohamud_Muse_Hersi             12  2017   

          month info_parenth             info_1 info_2                 info_3  \
61915      June         None      British actor  74–75                   None   
86907  February          (–)  Somali politician  79–80  President of Puntland   

      info_4 info_5 info_6 info_7 info_8 info_9 info_10 info_11  age  
61915   None   None   None   None   None   None    None    None  NaN  
86907   None   None   None   None   None   None    None    None  NaN

<IPython.core.display.Javascript object>

In [14]:
# For loop to find rows with values and pattern and calculate and extract age to age column and remove age from info_2
for i in df.loc[rows_to_check, :].index:
    item = df.loc[i, "info_2"]
    match = re.search(pattern, item)
    if match:
        age = (int(match.group(1)) + int(match.group(3))) / 2
        df.loc[i, "age"] = age
        df.loc[i, "info_2"] = re.sub(pattern, "", df.loc[i, "info_2"]).strip()

# Checking example rows
pd.concat([df[df["name"] == "Mohamud Muse Hersi"], df[df["name"] == "Basil Soper"]])

day                name  \
86907   8  Mohamud Muse Hersi   
61915   1         Basil Soper   

                                                      info  \
86907  , Somali politician, 79–80, President of Puntland .   
61915                              , British actor, 74–75.   

                                                   link num_references  year  \
86907  https://en.wikipedia.org/wiki/Mohamud_Muse_Hersi             12  2017   
61915         https://en.wikipedia.org/wiki/Basil_Soper              0  2013   

          month info_parenth             info_1 info_2                 info_3  \
86907  February          (–)  Somali politician         President of Puntland   
61915      June         None      British actor                          None   

      info_4 info_5 info_6 info_7 info_8 info_9 info_10 info_11   age  
86907   None   None   None   None   None   None    None    None  79.5  
61915   None   None   None   None   None   None    None    None  74.5

<IPython.core.display.Javascript object>

#### Extracting `age` as Single Integer

In [15]:
# List of patterns for age formats with single integer for age
pattern = r"\b(\d{1,3})\b"

# List and number of rows matching patterns
rows_to_check = rows_with_pattern(
    df[(df["age"].isna()) & (df["info_2"].notna())], "info_2", pattern
)

There are 24 rows with matching pattern in column 'info_2'.


<IPython.core.display.Javascript object>

In [16]:
# For loop to extract age pattern to age column
for i in df.loc[rows_to_check, :].index:
    item = df.loc[i, "info_2"]
    match = re.search(pattern, item)
    if match:
        age = int(match.group(1))
        df.loc[i, "age"] = age
        df.loc[i, "info_2"] = re.sub(pattern, "", df.loc[i, "info_2"]).strip()

# Re-checking number of rows matching patterns
rows_to_check = rows_with_pattern(
    df[(df["age"].isna()) & (df["info_2"].notna())], "info_2", pattern
)

There are 0 rows with matching pattern in column 'info_2'.


<IPython.core.display.Javascript object>

In [17]:
# Checking number of remaining missing values
print(f'There are {df["age"].isna().sum()} missing values for age.')
df[df["age"].isna()].sample(2)

There are 36 missing values for age.


day              name  \
118488  23  Makhosi Vilakati   
119194  12     Tohami Khaled   

                                                                              info  \
118488      , Swazi politician, minister of labour and social security , COVID-19.   
119194  , Libyan military officer, head of the Internal Security Agency, COVID-19.   

                                                  link num_references  year  \
118488  https://en.wikipedia.org/wiki/Makhosi_Vilakati              2  2021   
119194     https://en.wikipedia.org/wiki/Tohami_Khaled              9  2021   

           month info_parenth                   info_1  \
118488   January     (since )         Swazi politician   
119194  February         None  Libyan military officer   

                                        info_2    info_3 info_4 info_5 info_6  \
118488  minister of labour and social security  COVID-19   None   None   None   
119194    head of the Internal Security Agency  COVID-19   None   None   None   

       info_7 info_8 info_9 info_10 info_11  age  
118488   None   None   None    None    None  NaN  
119194   None   None   None    None    None  NaN

<IPython.core.display.Javascript object>

#### Observations:
- At this point, we could look through the list to manually extract any remaining `age` info, which would likely be the fastest approach.
- We will instead take a programmatic approach, for the sake of the exercise.
- `info_parenth`, and `info_3` and beyond are the remaining columns to search.
- We see that COVID-19 appears often and the number 19 could be mistakenly extracted as an age.  
- Let us start by moving it to a new column `cause_of_death`.

#### Extracting "COVID-19" from Remaining `info` Sub-columns where `age` Missing

In [18]:
# List of columns to check
cols_to_check = [
    "info_parenth",
    "info_3",
    "info_4",
    "info_5",
    "info_6",
    "info_7",
    "info_8",
    "info_9",
    "info_10",
    "info_11",
]

# Pattern for re
pattern = r"(COVID-19)"

# For loop to collect indices of all rows with pattern
comb_rows_to_check = []
for column in cols_to_check:
    rows_to_check = rows_with_pattern(
        df[(df["age"].isna()) & (df[column].notna())], column, pattern
    )
    comb_rows_to_check += rows_to_check

# Checking sample of rows
df.loc[comb_rows_to_check, :].sample(2)

There are 1 rows with matching pattern in column 'info_parenth'.
There are 30 rows with matching pattern in column 'info_3'.
There are 2 rows with matching pattern in column 'info_4'.
There are 1 rows with matching pattern in column 'info_5'.
There are 0 rows with matching pattern in column 'info_6'.
There are 0 rows with matching pattern in column 'info_7'.
There are 0 rows with matching pattern in column 'info_8'.
There are 0 rows with matching pattern in column 'info_9'.
There are 0 rows with matching pattern in column 'info_10'.
There are 0 rows with matching pattern in column 'info_11'.


day           name  \
116845   7        Tasiman   
119194  12  Tohami Khaled   

                                                                              info  \
116845                         , Indonesian politician, regent of Pati , COVID-19.   
119194  , Libyan military officer, head of the Internal Security Agency, COVID-19.   

                                               link num_references  year  \
116845        https://en.wikipedia.org/wiki/Tasiman              1  2020   
119194  https://en.wikipedia.org/wiki/Tohami_Khaled              9  2021   

           month info_parenth                   info_1  \
116845  December          (–)    Indonesian politician   
119194  February         None  Libyan military officer   

                                      info_2    info_3 info_4 info_5 info_6  \
116845                        regent of Pati  COVID-19   None   None   None   
119194  head of the Internal Security Agency  COVID-19   None   None   None   

       info_7 info_8 info_9 info_10 info_11  age  
116845   None   None   None    None    None  NaN  
119194   None   None   None    None    None  NaN

<IPython.core.display.Javascript object>

In [19]:
# For loop to extract COVID-19 from remaining info columns for entries with missing age
for column in cols_to_check:
    for i in df.loc[comb_rows_to_check, :].index:
        if df.loc[i, column]:
            item = df.loc[i, column]
            match = re.search(pattern, item)
            if match:
                cause = match.group(1)
                df.loc[i, "cause_of_death"] = cause
                df.loc[i, column] = re.sub(pattern, "", df.loc[i, column]).strip()

# Checking sample of rows
df.loc[comb_rows_to_check, :].sample(2)

day                    name  \
123455   2  Mohamed Nejib Berriche   
117861   7      Munira Yamin Satti   

                                                  info  \
123455       , Tunisian politician, deputy , COVID-19.   
117861  , Pakistani politician, Punjab MPA , COVID-19.   

                                                        link num_references  \
123455  https://en.wikipedia.org/wiki/Mohamed_Nejib_Berriche              3   
117861      https://en.wikipedia.org/wiki/Munira_Yamin_Satti              3   

        year    month info_parenth                info_1      info_2 info_3  \
123455  2021     July          (–)   Tunisian politician      deputy          
117861  2021  January     (since )  Pakistani politician  Punjab MPA          

       info_4 info_5 info_6 info_7 info_8 info_9 info_10 info_11  age  \
123455   None   None   None   None   None   None    None    None  NaN   
117861   None   None   None   None   None   None    None    None  NaN   

       cause_of_death  
123455       COVID-19  
117861       COVID-19

<IPython.core.display.Javascript object>

#### Observations:
- With "COVID-19" put aside, we can check for any remaining digits in the same columns for these entries.

#### Checking Remaining `info` Columns for Remaining Digits

In [20]:
# List of columns to check
cols_to_check = [
    "info_parenth",
    "info_3",
    "info_4",
    "info_5",
    "info_6",
    "info_7",
    "info_8",
    "info_9",
    "info_10",
    "info_11",
]

# Pattern for re
pattern = r"\d"

# For loop to collect indices of all rows with pattern
comb_rows_to_check = []
for column in cols_to_check:
    rows_to_check = rows_with_pattern(
        df[(df["age"].isna()) & (df[column].notna())], column, pattern
    )
    comb_rows_to_check += rows_to_check

# Checking sample of rows
df.loc[comb_rows_to_check, :]

There are 2 rows with matching pattern in column 'info_parenth'.
There are 0 rows with matching pattern in column 'info_3'.
There are 0 rows with matching pattern in column 'info_4'.
There are 0 rows with matching pattern in column 'info_5'.
There are 0 rows with matching pattern in column 'info_6'.
There are 0 rows with matching pattern in column 'info_7'.
There are 0 rows with matching pattern in column 'info_8'.
There are 0 rows with matching pattern in column 'info_9'.
There are 0 rows with matching pattern in column 'info_10'.
There are 0 rows with matching pattern in column 'info_11'.


day         name                       info  \
22855   10   Little Eva                        , .   
126710  31  Simon Young  , Irish radio presenter .   

                                                         link num_references  \
22855                https://en.wikipedia.org/wiki/Little_Eva             14   
126710  https://en.wikipedia.org/wiki/Simon_Young_(presenter)              9   

        year    month                                          info_parenth  \
22855   2003    April  (née Eva Narcissus Boyd), 59, American pop singer ()   
126710  2021  October                                             (RTÉ 2fm)   

                       info_1 info_2 info_3 info_4 info_5 info_6 info_7  \
22855                           None   None   None   None   None   None   
126710  Irish radio presenter   None   None   None   None   None   None   

       info_8 info_9 info_10 info_11  age cause_of_death  
22855    None   None    None    None  NaN            NaN  
126710   None   None    None    None  NaN            NaN

<IPython.core.display.Javascript object>

#### Observations:
- There are only two entries that still potentially contain digits for the age data.
- Here, we see that there is one entry we can preserve that has an age value in `info_parenth`.
- The other entry has a radio station identification value and is missing age data.
- After we collect this last age, we will drop the remaining entries missing `age`.

#### Extracting `age` from `info_parenth`

In [21]:
# List of patterns for age formats with single integer for age
pattern = r"\b(\d{1,3})\b"

# List and number of rows matching patterns
rows_to_check = rows_with_pattern(
    df[(df["age"].isna()) & (df["info_parenth"].notna())], "info_parenth", pattern
)

There are 1 rows with matching pattern in column 'info_parenth'.


<IPython.core.display.Javascript object>

In [22]:
# For loop to extract age pattern to age column
for i in df.loc[rows_to_check, :].index:
    item = df.loc[i, "info_parenth"]
    match = re.search(pattern, item)
    if match:
        age = int(match.group(1))
        df.loc[i, "age"] = age
        df.loc[i, "info_parenth"] = re.sub(
            pattern, "", df.loc[i, "info_parenth"]
        ).strip()

# Re-checking number of rows matching patterns
rows_to_check = rows_with_pattern(
    df[(df["age"].isna()) & (df["info_parenth"].notna())], "info_parenth", pattern
)

There are 0 rows with matching pattern in column 'info_parenth'.


<IPython.core.display.Javascript object>

#### Dropping the Last Entries with Missing `age` Values

In [23]:
# Checking number of remaining missing values
print(f'There are {df["age"].isna().sum()} missing values for age.')

There are 35 missing values for age.


<IPython.core.display.Javascript object>

In [24]:
# Dropping rows, resetting index, and checking new shape of df
df.dropna(subset="age", inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132393, 21)

<IPython.core.display.Javascript object>

In [25]:
# Checking current info status
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132393 entries, 0 to 132392
Data columns (total 21 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   day             132393 non-null  object 
 1   name            132393 non-null  object 
 2   info            132393 non-null  object 
 3   link            132393 non-null  object 
 4   num_references  132393 non-null  object 
 5   year            132393 non-null  int64  
 6   month           132393 non-null  object 
 7   info_parenth    49752 non-null   object 
 8   info_1          132393 non-null  object 
 9   info_2          132371 non-null  object 
 10  info_3          62537 non-null   object 
 11  info_4          12584 non-null   object 
 12  info_5          1504 non-null    object 
 13  info_6          217 non-null     object 
 14  info_7          33 non-null      object 
 15  info_8          7 non-null       object 
 16  info_9          1 non-null       object 
 17  info_10   

<IPython.core.display.Javascript object>

#### Observations:
- We have 132,393 entries containing the target variable `age`.
- Some of these rows may represent groups or members of non-human species, as we have observed previously.
- We have been replacing values extracted with empty strings.  Before moving forward let us replace these empty strings with Nan, as it will simplify slicing the dataframe.
- Then it will be time to search for nationality.

#### Replacing Empty Strings with NaN

In [26]:
# Replacing empty strings with NaN
df = df.replace(r"^\s*$", np.nan, regex=True)

<IPython.core.display.Javascript object>

In [27]:
# Checking the NaN values per column
df.isna().sum()

day                    0
name                   0
info                   0
link                   0
num_references         0
year                   0
month                  0
info_parenth       82641
info_1            132363
info_2                48
info_3             70065
info_4            119852
info_5            130895
info_6            132179
info_7            132362
info_8            132387
info_9            132392
info_10           132392
info_11           132392
age                    0
cause_of_death    132393
dtype: int64

<IPython.core.display.Javascript object>

## Extracting Nationality Data

In [28]:
# Checking a sample
df.sample(5)

day                name  \
41152  15  Natalya Estemirova   
17259  29       Frances Bible   
53469  16      Jacques Forest   
88705   8         Peter Vogel   
15879  19   Christiane Herzog   

                                                                        info  \
41152  , 50, Russian human rights activist in Chechnya and Ingushetia, shot.   
17259                                , 82, American operatic mezzo-soprano .   
53469                                            , 91, French carcinologist.   
88705                                                   , 80, German artist.   
15879       , 63, wife of Roman Herzog, former President of Germany, cancer.   

                                                     link num_references  \
41152    https://en.wikipedia.org/wiki/Natalya_Estemirova             33   
17259         https://en.wikipedia.org/wiki/Frances_Bible              5   
53469        https://en.wikipedia.org/wiki/Jacques_Forest              6   
88705  https://en.wikipedia.org/wiki/Peter_Vogel_(artist)              7   
15879     https://en.wikipedia.org/wiki/Christiane_Herzog              6   

       year     month           info_parenth info_1  \
41152  2009      July                   None    NaN   
17259  2001   January  (New York City Opera)    NaN   
53469  2012  February                   None    NaN   
88705  2017       May                   None    NaN   
15879  2000      June                   None    NaN   

                                                         info_2  \
41152  Russian human rights activist in Chechnya and Ingushetia   
17259                           American operatic mezzo-soprano   
53469                                      French carcinologist   
88705                                             German artist   
15879                                      wife of Roman Herzog   

                            info_3  info_4 info_5 info_6 info_7 info_8 info_9  \
41152                         shot    None   None   None   None   None   None   
17259                         None    None   None   None   None   None   None   
53469                         None    None   None   None   None   None   None   
88705                         None    None   None   None   None   None   None   
15879  former President of Germany  cancer   None   None   None   None   None   

      info_10 info_11   age  cause_of_death  
41152    None    None  50.0             NaN  
17259    None    None  82.0             NaN  
53469    None    None  91.0             NaN  
88705    None    None  80.0             NaN  
15879    None    None  63.0             NaN

<IPython.core.display.Javascript object>

#### Observations:
- `info_2` appears overall consistent with the Wikipedia field  that combines "citizenship" and "known for".
- The first word does appear to represent the nationality.
- Recall that this information is in `info_1` for some entries and may also be in other `info` columns beyond `info_2`.
- Running the sample check a few times reveals that when citizenship changed the original citizenship may be followed by '-born', then the second citizenship.
- There are nationalities that have multiple words and there are capitalized words that are not part of the nationality.
- We will start with a list of nationalities downloaded from marijn's github repository [List of nationalities](https://gist.github.com/marijn/274449).

### Reading List of Nationalities

In [29]:
# Reading in list of nationalities
nationalities = pd.read_csv(
    "nationalities.txt", sep="/n", engine="python", names=["Nationality"]
)

print(nationalities.shape)
nationalities.head()

(194, 1)


Nationality
0      Afghan
1    Albanian
2    Algerian
3    American
4    Andorran

<IPython.core.display.Javascript object>

#### Observations:
- There are two nationalities in the current list that might present differently in our dataset:  'Kittian and Nevisian' and 'Trinidadian or Tobagonian', so we will add their individual parts to the list.
- As we encounter other versions for nationality values, we can continue to update the list.

In [30]:
# Converting nationalities to a list
nationalities_lst = nationalities["Nationality"].to_list()
nationalities_lst

['Afghan',
 'Albanian',
 'Algerian',
 'American',
 'Andorran',
 'Angolan',
 'Antiguans',
 'Argentinean',
 'Armenian',
 'Australian',
 'Austrian',
 'Azerbaijani',
 'Bahamian',
 'Bahraini',
 'Bangladeshi',
 'Barbadian',
 'Barbudans',
 'Batswana',
 'Belarusian',
 'Belgian',
 'Belizean',
 'Beninese',
 'Bhutanese',
 'Bolivian',
 'Bosnian',
 'Brazilian',
 'British',
 'Bruneian',
 'Bulgarian',
 'Burkinabe',
 'Burmese',
 'Burundian',
 'Cambodian',
 'Cameroonian',
 'Canadian',
 'Cape Verdean',
 'Central African',
 'Chadian',
 'Chilean',
 'Chinese',
 'Colombian',
 'Comoran',
 'Congolese',
 'Costa Rican',
 'Croatian',
 'Cuban',
 'Cypriot',
 'Czech',
 'Danish',
 'Djibouti',
 'Dominican',
 'Dutch',
 'East Timorese',
 'Ecuadorean',
 'Egyptian',
 'Emirian',
 'Equatorial Guinean',
 'Eritrean',
 'Estonian',
 'Ethiopian',
 'Fijian',
 'Filipino',
 'Finnish',
 'French',
 'Gabonese',
 'Gambian',
 'Georgian',
 'German',
 'Ghanaian',
 'Greek',
 'Grenadian',
 'Guatemalan',
 'Guinea-Bissauan',
 'Guinean',
 'Gu

<IPython.core.display.Javascript object>

In [31]:
# Adding nationality vaulues to nationality list
nationalities_lst += ["Kittian", "Nevisian", "Trinidadian", "Tobagonian"]

<IPython.core.display.Javascript object>

#### Observations:
- There are now 196 nationalities represented in the nationalities list, which we may use for comparison.
- There are single nationalities and dual nationalities represented ('Nationality1-born Nationality2' or "Nationality1-Nationality2").
- We will take the approach of extracting values that contain two nationalities, then address the remaining missing values.

### Extracting Nationality from `info_1` to `nation_1` and `nation_2`

#### Values with 2 Nationalities containing '-born'

In [106]:
test = df["info_1"].unique()
test = list(test)
test = test[1:]
test = test + ["Irish-born American Indian"]
test

['American politician',
 'Swedish Olympic sprinter',
 'American gridiron football player',
 'Colombian writer',
 'British businessman',
 'Austrian-British social psychologist',
 'King of Nepal',
 'New Zealand Maori leader',
 'American artist',
 'English sports journalist',
 'Chilean artist',
 'Jules Engel',
 'aka',
 'Jr',
 'professional wrestler',
 'British automotive engineer',
 'American materials scientist',
 'Australian weightlifter',
 'Icelandic Olympic athlete',
 'British actor',
 'American Olympic gymnast',
 'Somali politician',
 'Scottish broadcaster and writer',
 'French Olympic swimmer',
 'French Olympic boxer',
 'German Olympic wrestler',
 'Austrian Olympic sailor',
 'American basketball player',
 'American college basketball coach',
 'German choral conductor',
 'Irish-born American Indian']

<IPython.core.display.Javascript object>

In [117]:
nation_1_test = []
nation_2_test = []
role_test = []
missing = []
for item in test:
    if "-born" in item:
        split_1 = item.split("-born")
        nation_1 = split_1[0].strip()
        if nation_1 in nationalities_lst:
            nation_1_test.append(nation_1)

        split_nation = split_1[1].split()
        nation_2 = ""
        for word in split_nation:
            if word[0].isupper():
                nation_2 = (nation_2 + " " + word).strip()
                if nation_2 in nationalities_lst:
                    nation_2_test.append(nation_2)
                else:
                    missing.append(nation_2)

        role = ""
        for word in split_nation:
            if word[0].islower():
                role = role + " " + word
        role_test.append(role.strip())

    elif "-" in item:
        split_1 = item.split("-")
        nation_1 = split_1[0].strip()
        if nation_1 in nationalities_lst:
            nation_1_test.append(nation_1)
        else:
            missing.append(nation_1)

        split_nation = split_1[1].split()
        nation_2 = ""
        for word in split_nation:
            if word[0].isupper():
                nation_2 = (nation_2 + " " + word).strip()
                if nation_2 in nationalities_lst:
                    nation_2_test.append(nation_2)
                else:
                    missing.append(nation_2)

        role = ""
        for word in split_nation:
            if word[0].islower():
                role = role + " " + word
        role_test.append(role.strip())
    else:
        split_nation = item.split()
        nation_1 = ""
        for word in split_nation:
            if word[0].isupper():
                nation_1 = (nation_1 + " " + word).strip()
                if nation_1 in nationalities_lst:
                    nation_1_test.append(nation_1)

        role = ""
        for word in split_nation:
            if word[0].islower():
                role = role + " " + word
        role_test.append(role.strip())

nation_1_test
nation_2_test
role_test
missing

['American Indian']

<IPython.core.display.Javascript object>

In [ ]:
# List of columns to check
cols_to_check = [
    "info_1",
    "info_2",
    "info_3",
    "info_4",
    "info_5",
    "info_6",
    "info_7",
    "info_8",
    "info_9",
    "info_10",
    "info_11",
]

# List and number of rows matching patterns

    
comb_rows_to_check = []
for column in cols_to_check:
    for nation in nationalities_lst:
        # Pattern for re
        pattern = r"([A-Z][a-z]*\s?[A-Z]?[a-z]*\s?[A-Z]?[a-z]*)-born\s(%s)" % nation
        rows_to_check = rows_with_pattern(df[df[column].notna()], column, pattern)
        comb_rows_to_check += rows_to_check

# Checking a sample of rows
df.loc[comb_rows_to_check, :].sample(2)

#### Observations:
- Such values are only in `info_2` and `info_3`, so we will treat those values as indicated, one column at a time.

#### `info_2`

In [ ]:
# Column to be treated
column = "info_2"

# Empty list to collect missing nationalities
missing = []

# For loop to extract nation information to new columns nation_1 and nation_2
for i in df[df[column].notna()].index:
    item = df.loc[i, column]
    match = re.search(pattern, item)

    if match:
        print(match.group(0))
#         nation_1 = match.group(1).strip()
#         nation_2 = match.group(2).strip()

#         if nation_1 in nationalities_lst:
#             df.loc[i, "nation_1"] = nation_1
#             df.loc[i, column] = df.loc[i, column].replace(nation_1, "").strip()

#             if nation_2 in nationalities_lst:
#                 df.loc[i, "nation_2"] = nation_2
#                 df.loc[i, column] = df.loc[i, column].replace(nation_2, "").strip()
#             else:
#                 missing.append(nation_2)
#         else:
#             missing.append(nation_1)

# # Re-check number of and example rows
# rows_to_check = rows_with_pattern(df[df[column].notna()], column, pattern)
# df[df["name"] == "Giltedge"]

In [ ]:
set(missing)

In [ ]:
nationalities_lst

#### `info_3`

In [ ]:
# List of columns to check
column = "info_3"

# Pattern for re
pattern = r"([A-Z][a-z]*-born\s[A-Z][a-z]*\s)"

# List and number of rows matching patterns
rows_to_check = rows_with_pattern(df[df[column].notna()], column, pattern)

# Checking a sample of rows
df.loc[rows_to_check, :]

In [ ]:
# Column to be treated
column = "info_3"

# Extracting nation information to new columns nation_1 and nation_2
for i in df[df[column].notna()].index:
    item = df.loc[i, column]
    match = re.search(pattern, item)
    if match:
        nations = match.group(1).split("-born")
        nation_1 = nations[0].strip()
        nation_2 = nations[1].strip()
        df.loc[i, "nation_1"] = nation_1
        df.loc[i, "nation_2"] = nation_2
        df.loc[i, column] = re.sub(pattern, "", df.loc[i, column]).strip()

# Re-check number of and example rows
rows_to_check = rows_with_pattern(df[df[column].notna()], column, pattern)
df[df["name"] == "Herbert Freudenberger"]

#### Values with 2 Nationalities Hyphenated

In [ ]:
# Column to check and treat
column = "info_2"

# Patter for re
pattern = r"(^[A-Z][a-z]*-[A-Z][a-z]*)"

# List and number of rows matching patterns
rows_to_check = rows_with_pattern(
    df[(df["nation_1"].isna()) & (df[column].notna())], column, pattern
)

df.loc[rows_to_check, :].sample(2)

#### Observations:
- This pattern may have matches that are not nationalities, so we will add that comparison to our for loop when extracting.

In [ ]:
# Extracting nation information to new columns nation_1 and nation_2
for i in df[(df[column].notna()) & (df["nation_1"].isna())].index:
    item = df.loc[i, column]
    match = re.search(pattern, item)
    not_in_list = []

    if match:
        nations = match.group(0).split("-")
        nation_1 = nations[0].strip()
        nation_2 = nations[1].strip()
        if nation_1 in nationalities_lst:
            df.loc[i, "nation_1"] = nation_1
            df.loc[i, column] = df.loc[i, column].replace(nation_1, "").strip()
        else:
            not_in_list.append(nation_1)

        if nation_2 in nationalities_lst:
            df.loc[i, "nation_1"] = nation_2
            df.loc[i, column] = df.loc[i, column].replace(nation_2, "").strip()
        else:
            not_in_list.append(nation_2)

# Re-check number of and example rows
rows_to_check = rows_with_pattern(
    df[(df["nation_1"].isna()) & (df[column].notna())], column, pattern
)
df.loc[rows_to_check, :]

In [ ]:
df["nation_1"].unique()

In [ ]:
nationalities_lst

In [ ]:
# For loop to extract nationality from info_1 to nation_1 and nation_2
for i in df[df["info_1"].notna()].index:
    item = df.loc[i, "info_1"]

    nations = []
    for nationality in nationalities_lst:
        if nationality in item:
            nations.append(nationality)
            print(nations)

#     for j, nation in enumerate(nations):
#         df.loc[i, f"nation_{j+1}"] = nation
#         df.loc[i, "info_1"] = df.loc[i, "info_1"].replace(nation, "")

# # Checking sample of treated rows
# df[df["nation_1"].notna()].sample(2)

#### Observations:
- We have been fortunate in `info_1` in that it has provided a small sample on which to test our code and see a subset of values.
- Let us check the remaining unique values in `info_1` for any missed nationalities.

#### Checking Remaining Unique Values in `info_1`

In [ ]:
# Checking unique values in info_1
df["info_1"].unique()

#### Observations:
- `info_1` had only single nationality values, so we do not yet have a `nation_2` column.
- It appears "New Zealand Maori", "English", and "Icelandic" are not included in the nationalities list, but they are used as values on Wikipedia.
- We will take the approach of adding them to the list and re-running the code to extract values from `info_1`.

#### Adding Missing Nationalities to `nationalities_lst` and Repeating Extraction of Nationality from `info_1`

In [ ]:
# Confirming nationalities are missing then adding them to existing nationalities list
missing_nations = ["New Zealand Maori", "English", "Icelandic"]
missing_nations = [item for item in missing_nations if item not in nationalities_lst]
nationalities_lst += missing_nations

# List of new columns for first and second nationality
new_cols = ["nation_1", "nation_2"]

# For loop to extract nationality from info_1 to nation_1 and nation_2
for col in new_cols:

    for i in df[df["info_1"].notna()].index:
        item = df.loc[i, "info_1"]

        for nationality in nationalities_lst:
            if nationality in item:
                df.loc[i, col] = nationality
                df.loc[i, "info_1"] = df.loc[i, "info_1"].replace(nationality, "")

# Checking sample of treated rows
df[df["nation_1"].notna()].sample(2)

#### Re-checking Remaning Unique Values in `info_1`

In [ ]:
# Re-checking remaining unique values in info_1
df["info_1"].unique()

#### Observations:
- That iteration addressed missing nationalities.
- We do have nationality given as a country, "Nepal", that we will need to keep in mind.

### Extracting Nationality from `info_2` to `nation_1` and `nation_2`

In [ ]:
# List of new columns for first and second nationality
new_cols = ["nation_1", "nation_2"]

# For loop to extract nationality from info_2 to nation_1 and nation_2
for col in new_cols:

    for i in df[df["info_2"].notna()].index:
        item = df.loc[i, "info_2"]

        for nationality in nationalities_lst:
            if nationality in item:
                df.loc[i, col] = nationality
                df.loc[i, "info_2"] = df.loc[i, "info_2"].replace(nationality, "")

# Checking sample of treated rows
df[df["nation_1"].notna()].sample(2)

In [ ]:
df["info_2"].unique()

### Exporting Dataset to SQLite Database [wp_life_expect_clean2.db](https://github.com/teresahanak/wikipedia-life-expectancy/blob/main/wp_life_expect_clean1.db)

In [ ]:
# Saving complete raw dataset in a SQLite database
conn = sql.connect("wp_life_expect_clean2.db")
df.to_sql("wp_life_expect_clean2", conn, index=False)

# [Proceed to Notebook 4 of  4:  Data Cleaning Part 3](https://github.com/teresahanak/wikipedia-life-expectancy/blob/main/wp_life_expect_data_clean2_thanak_2022_06_17.ipynb)